aptos-move/block-executor/src/scheduler_status.rs

```diff
@@ -0,0 +1,1387 @@
+            if self.num_stalls.load(Ordering::Relaxed) > 0 {
+                return Ok(false);
+            }
+
+            if let Some(incarnation) = inner_status.pending_scheduling() {
+                if incarnation == 0 {
+                    // Invariant due to scheduler logic: for a successful remove_stall there
+                    // must have been an add_stall for incarnation 0, which is impossible.
+                    return Err(code_invariant_error("0-th incarnation in remove_stall"));
+                }
+                self.execution_queue_manager
+                    .add_to_schedule(incarnation == 1, self.txn_idx);
+            } else if inner_status.is_executed() {
+                // TODO(BlockSMTv2): Here, when waiting is supported, if inner status is executed,
+                // would need to notify waiting workers.
+
+                // Status is Executed so the dependency shortcut flag may not be
+                // EXECUTING (finish_execution sets Executed status and DEFER or SAFE flag).
+                self.swap_dependency_flag_any(
+                    &[dependency_flags::DEFER, dependency_flags::SAFE],
+                    dependency_flags::SAFE,
+                    "remove_stall",
+                )?;
+            }
+
+            return Ok(true);
+        }
+        Ok(false)
+    }
+
+    /// Checks if an incarnation has already been marked for abortion.
+    ///
+    /// This can be called during an ongoing execution to determine if the
+    /// execution has been concurrently aborted. This allows the executor
+    /// to return early and to discard the results.
+    pub(crate) fn already_try_aborted(&self, incarnation: Incarnation) -> bool {
+        self.next_incarnation_to_try_abort.load(Ordering::Relaxed) > incarnation
+    }
+
+    /// Checks that the shortcut flag is SAFE. This is used by the scheduler
+    /// when removing a previously propagated stall signal, and it is safe to
+    /// use the shortcutbecause of the best-effort nature of the stall mechanism.
+    pub(crate) fn shortcut_executed_and_not_stalled(&self) -> bool {
+        matches!(
+            self.dependency_shortcut.load(Ordering::Relaxed),
+            dependency_flags::SAFE
+        )
+    }
+
+    /// Checks if the transaction is ready for scheduling and not stalled.
+    /// This is used to determine if a transaction should be considered
+    /// for execution by the scheduler.
+    pub(crate) fn pending_scheduling_and_not_stalled(&self) -> bool {
+        let guard = self.inner_status.lock();
+        guard.pending_scheduling().is_some() && self.num_stalls.load(Ordering::Relaxed) == 0
+    }
+
+    // === More inner status wrappers ===
+
+    pub(crate) fn is_executed(&self) -> bool {
+        self.inner_status.lock().is_executed()
+    }
+
+    pub(crate) fn ever_executed(&self) -> bool {
+        self.inner_status.lock().ever_executed()
+    }
+
+    pub(crate) fn last_incarnation(&self) -> Incarnation {
```

Comment:
last -> current?

Reply:
maybe just incarnation() even. It's used to return commit tasks in the scheduler (coming next)